Introduce DFG based combinational logic optimizer

gezalore opened the pull request:
I am raising this against V5, for a few reasons
- The scheduling improvements on V5 make the MT results a bit more stable (the fixed point of the peephole pass depends on the order you visit vertices, and can perturb the logic enough to see big differences in MT scheduling depending on iteration order on master, which is absent with the V5 scheduler - I believe this is due to a fix present on V5 for merging independent MTasks)
- As a bit of motivation for the user to upgrade
- Potentially a bug farm, though I have not seen issues for a while

There are improvements that could be made about the implementation (for Verialion performance in particular), but I would like to punt those to a later patch, it is not bad as is.

Still todo for this patch:
- [x] Add brief section to internals docs
- [x] Re-check the usefulness of some of the peephole patterns
- [x] Limit the extract pass to combinational logic only

I will post a few questions in code shortly.

---

Added a new data-flow graph (DFG) based combinational logic optimizer.
The capabilities of this covers a combination of V3Const and V3Gate, but
is also more capable of transforming combinational logic into simplified
forms and more.

This entail adding a new internal representation, `DfgGraph`, and
apporpriate `astToDfg` and `dfgToAst` conversion functions. The graph
represents some of the combinational equations (~continuous assignments)
in a module, and for the duration of the DFG passes, it takes over the
role of AstModule. A bulk of the Dfg vertices represent expressions.
These vertex classes, and the corresponding conversions to/from AST are
mostly auto-generated by astgen, together with a DfgVVisitor that can be
used for dynamic dispatch based on vertex (operation) types.

The resulting combinational logic graph (a `DfgGraph`) is then optimized
in various ways. Currently we perform common sub-expression elimination,
variable inlining, and some specific peephole optimizations, but there
is scope for more optimizations in the future using the same
representation. The optimizer is run directly before and after inlining.
The pre inline pass can operate on smaller graphs and hence converges
faster, but still has a chance of substantially reducing the size of the
logic on some designs, making inlining both faster and less memory
intensive. The post inline pass can then optimize across the inlined
module boundaries. No optimization is performed across a module
boundary.

For debugging purposes, each peephole optimization can be disabled
individually by setting an environment variable to 0:
VERILATOR_DFG_PEEPHOLE_ENABLE__<OPT>=0 where <OPT> is one of the
optimizations listed in V3DfgPeephole.h, for example
VERILATOR_DFG_PEEPHOLE_ENABLE__REMOVE_NOT_NOT=0.

The DFG optimizer can also be disabled via the opt_dfg_{off,on}
metacomments in source, or the equivalent configuration file directives.
The purpose of this is for debugging and testing only, and the user is
not expected to use these in production.

The peephole patterns currently implemented were mostly picked based on
the design that inspired this work, and on that design the optimizations
yields ~30% single threaded speedup, and ~50% speedup on 4 threads. As
you can imagine not having to haul around redundant combinational
networks in the rest of the compilation pipeline also helps with memory
consumption, and up to 30% peak memory usage of Verilator was observed
on the same design.

Gains on other arbitrary designs are smaller (and can be improved by
analyzing those designs). For example OpenTitan gains between 1-8%
speedup depending on build type.

gezalore commented:
If you are curious about instruction counts, here is some data on XiangShan (the design that inspired this) single threaded (these are on top of master actually, but V5 is similar). Also note that this design is generated with Chisel, so there are a lot of things we can improve on, so neat hand written logic will benefit less at the moment (but I still think we can do better using the same representation):

Baseline:
``` 
 1,398,283,670,861      cycles                    #    2.594 GHz
   959,264,494,613      instructions              #    0.69  insn per cycle
    64,153,575,807      branches                  #  119.009 M/sec
    10,769,400,379      branch-misses             #   16.79% of all branches

          539.156545584 seconds time elapsed
```

With DFG:
``` 
 1,042,557,870,449      cycles                    #    2.594 GHz
   788,250,011,995      instructions              #    0.76  insn per cycle
    53,959,296,894      branches                  #  134.254 M/sec
     6,369,035,323      branch-misses             #   11.80% of all branches

          401.981852909 seconds time elapsed
```

gezalore commented on src/V3Dataflow.cpp:
```diff
@@ -0,0 +1,291 @@
+#include <vector>
+
+// Extract more combinational logic equations from procedures for better optimization opportunities
+class DataflowExtractVisitor final : public VNVisitor {
```
I will restrict this to work on combinational logic (`always_comb`/`always @*` only, as otherwise we might do more work than needed with sparse clock events.

wsnyder replied:
Ok

gezalore replied:
Done

gezalore commented on src/V3DfgAstToDfg.cpp:
```diff
@@ -0,0 +1,239 @@
+    return new Vertex{dfg, nodep->fileline(), static_cast<uint32_t>(nodep->width())};
+}
+
+// Currently unhandled nodes
+template <>  //
+DfgAtoN* makeVertex<DfgAtoN, AstAtoN>(const AstAtoN*, DfgGraph&) {
+    return nullptr;
+}
+
+template <>  //
+DfgCCast* makeVertex<DfgCCast, AstCCast>(const AstCCast*, DfgGraph&) {
+    return nullptr;
+}
+
+template <>  //
+DfgCompareNN* makeVertex<DfgCompareNN, AstCompareNN>(const AstCompareNN*, DfgGraph&) {
+    return nullptr;
+}
+
+template <>  //
+DfgSliceSel* makeVertex<DfgSliceSel, AstSliceSel>(const AstSliceSel*, DfgGraph&) {
+    return nullptr;
+}
+
+template <>  // TODO: Select into final dimension should work fine
+DfgArraySel* makeVertex<DfgArraySel, AstArraySel>(const AstArraySel*, DfgGraph&) {
+    return nullptr;
+}
```
Anything that returns a nullptr here is left as AST and is not converted to DFG (unhandled), can you think of any more AstNodeMath classes that might be problematic and should be included here?

wsnyder replied:
Why are compares and casts problematic?   I'd think this should generally be anything imprure, and arrays as your code doesn't handle them.  What about function and member calls?

gezalore replied:
I turned off casts because they change dtypes/widths in some non-trivial (for me) way, but they should not really exist at all at the point this is run, but being paranoid.

CompareNN operates on strings. In general should probably turn all this off on string (they don't really have a width), and float/double, I will do that.

Calls are not expressions (not AstNodeMath) so are currently covered by the default case here:
https://github.com/gezalore/verilator/blob/f76e701919462d2a1060742c12408d73f9880a93/src/V3DfgAstToDfg.cpp#L119-L123

I think I am missing the impure check, will add that.

wsnyder replied:
Yes, better to disable for a string dtype, as opposed string operators.

gezalore replied:
I actually went the opposite direction and support only explicitly bit-vector like basic types and packed arrays of the same for now. On XiangShan I get the same results. We can add handling for more complex types later.

gezalore commented on src/V3DfgAstToDfg.cpp:
```diff
@@ -0,0 +1,239 @@
+        if (nodep->user2()) getNet(nodep)->setHasExtRefs();
+    }
+
+    void visit(AstAssignW* nodep) override {
```
Currently we only convert AssignW. We should also do `always_comb` etc that only has a single assignment at least. Preferably future patch.

gezalore commented on src/V3DfgAstToDfg.cpp:
```diff
@@ -0,0 +1,239 @@
+            return;
+        }
+
+        // Cannot handle assignments between different types at the moment
+        if (!nodep->lhsp()->dtypep()->skipRefp()->similarDType(
+                nodep->rhsp()->dtypep()->skipRefp())) {
+            markReferenced(nodep);
+            ++m_ctx.m_nonRepresentableDType;
+            return;
+        }
```
Is there a more robust way of detecting if the two sides are simple packed types (all I really care about is that they are both packed and have the same width, possibly signedness)? This for example fails if one side is `logic` but the other is `logic [0:0]` (which you get from parametric source), even though those two are fine here.

wsnyder replied:
Not currently, the type handling is messy.  Note "logic" and "logic [0]" while in your code and much of the code can be handled similarly they are not the same language wise, or if hits DPI or tracing.

gezalore commented on src/V3DfgDfgToAst.cpp:
```diff
@@ -0,0 +1,318 @@
+                // We found a DfgVar driven by this node
+                varp = getCanonicalVar(sinkDfgVarp);
+            } else {
+                // No DfgVar driven by this node. Create a temporary.
+                const string name = m_tmpNames.get(vtxp->hash(m_hashCache).toString());
+                // TODO: Do we ever need to make this VSigning::SIGNED?
+                AstNodeDType* const dtypep = v3Global.rootp()->findBitDType(
+                    vtxp->width(), vtxp->width(), VSigning::UNSIGNED);
+                varp = new AstVar{vtxp->fileline(), VVarType::MODULETEMP, name, dtypep};
+                // Add temporary AstVar to containing module
+                m_modp->addStmtp(varp);
```
I wonder if we ever need to make a signed temporary here?
This is used to render code that has common sub-expressions:
```systemverilog
  assign x = a + b - p;
  assign y = a + b - q;
```
At the end of DFG this will be rendered as 
```systemverilog
   wire logic [width-1:0] tmp = a + b;
   assign x = tmp + p;
   assign y = tmp + q;
```
So the question is, is it safe for tmp to always be unsigned?

I could restrict AstToDfg so anything with a signed term is not optimized, but there is also `$signed` and friends that might introduce more. It would be ideal to handle these.

wsnyder replied:
I think the temp should be able to keep the signing of the expression to avoid trouble, while I can't think of things that might break this is what all other temporary insertions do.  I don't see what would prevent optimizing signing - the rules should be setup to allow those cases - generally the Ast types are different for any cases where signing matters.

gezalore replied:
> generally the Ast types are different for any cases where signing matters

That is my understanding. So can I just set the temporary signed, if the DFG vertex that generates it is a signed operation?

wsnyder replied:
Correct, or generally just take the dtype entirely from the expression.

gezalore replied:
> or generally just take the dtype entirely from the expression.

Dfg vertices do not have dtypes, just widths, hence my peril :) (and would prefer not to add one as it's not really relevant, and hard/verbose to keep updaed, the vertex type should fully determine what the result is, and like AstNodeMath, the Dfg vertex type determines the interpretation of the input, otherwise everything is just a packed vector of bits)

gezalore replied:
I think I have reasoned myself into believing that always unsigned temporaries are OK, because we only add a temporary for intermediate expressions that are read by other expressions (i.e.: there will be no readers other than those expressions), and expressions (past V3Width) interpret their inputs based on the expression node type, not the operand types.

Does this sound reasonable?

(this also made me realise that we should not merge signed and unsigned *variables* in this pass, so I will fix that).

gezalore replied:
> (this also made me realise that we should not merge signed and unsigned variables in this pass, so I will fix that).

Looking more into this, even this should not matter post V3Width and some early front-end passes (basically where we need to determine whether to sign or zero extend soemthing, this optimization is after those). Signed and unsigned variables are emitted the same in the output and both have unused upper bits cleared by V3Clean on assignment, so again I think always unsigned temporaries here are OK.

gezalore commented on src/V3DfgDfgToAst.cpp:
```diff
@@ -0,0 +1,318 @@
+
+    // Predicate checking whether a variable should be retained,
+    // even if it's equal to another variable
+    static bool keep(const DfgVar* vtxp) {
+        // Keep if referenced outide this module
+        if (vtxp->hasExtRefs()) return true;
+        // Keep if traced
+        if (v3Global.opt.trace() && vtxp->varp()->isTrace()) return true;
+        // Otherwise it can be removed
+        return false;
+    }
```
Generally we throw away redundant variables. This tells us when not to. Can you think of any more cases? Coverage is inserted before DFG, so I think that should be fine.

wsnyder replied:
public.

gezalore replied:
Fixed

gezalore reviewed (commented):
@wsnyder could you look at the questions below please.

wsnyder reviewed (commented):
Some feedback, certainly will be more next round.

gezalore commented on src/V3Error.h:
```diff
@@ -48,6 +48,7 @@ class V3ErrorCode final {
         // Boolean information we track per-line, but aren't errors
         I_CELLDEFINE,   // Inside cell define from `celldefine/`endcelldefine
         I_COVERAGE,     // Coverage is on/off from /*verilator coverage_on/off*/
+        I_OPT_DFG,      // DFG optimizer is on/off from /*verilator opt_dfg_on/off*/
```
This is for debugging, to make it possible to bisect the input (via the .vlt config file) to locate logic that is mis-optimized (it's pretty hopeless to track it down otherwise in big designs). Do you have an alternative suggestion that would allow for that without this mechanism?

Also used by t_dfg_peephole for testing but I can change that.

BTW I agree this should not be something the user should have to use, unless asked explicitly for debugging, so should be considered a "hidden" option in any case.

gezalore replied:
Removed and rewritten tests that used this.

gezalore commented on src/V3DfgPeephole.h:
```diff
@@ -0,0 +1,108 @@
+
+// clang-format off
+#define FOR_EACH_DFG_PEEPHOLE_OPTIMIZATION(macro)   \
+    _FOR_EACH_DFG_PEEPHOLE_OPTIMIZATION_APPLY(macro, SWAP_CONST_IN_COMMUTATIVE_BINARY) \
+    _FOR_EACH_DFG_PEEPHOLE_OPTIMIZATION_APPLY(macro, SWAP_NOT_IN_COMMUTATIVE_BINARY)   \
+    _FOR_EACH_DFG_PEEPHOLE_OPTIMIZATION_APPLY(macro, SWAP_VAR_IN_COMMUTATIVE_BINARY) \
```
Not sure what you mean "prove these", I checked nodist/code_coverage andV3DfgPeephole.cpp was 100%-epsilon for some trivial omission so far.

gezalore commented:
>The scheduling improvements on V5 make the MT results a bit more stable (the fixed point of the peephole pass depends on the order you visit vertices, and can perturb the logic enough to see big differences in MT scheduling depending on iteration order on master, which is absent with the V5 scheduler - I believe this is due to a fix present on V5 for merging independent MTasks) 

Note to self: This conjecture tuns out to be incorrect. I manged to an MT schedule on v5 that was 20% slower than some others, just by fiddling with some optimizations that only effect a very small number of elements

gezalore commented on src/V3Ast.h:
```diff
@@ -97,7 +97,7 @@ class VNType final {
     inline VNType() = default;
     // cppcheck-suppress noExplicitConstructor
-    inline VNType(en _e)
+    constexpr inline VNType(en _e)
```
Will do before merging, leaving it here now in case there is more similar pre-pr stuff.

gezalore commented on docs/guide/exe_verilator.rst:
```diff
@@ -454,6 +463,18 @@ Summary:
+
+   Disable all use of the DFG based combinational logic optimizer
+
+.. option:: -fno-pre-inline-dfg
+
+   Do not apply the DFG optimizer before inlining
+
+.. option:: -fno-post-inline-dfg
```
Turns out those are the actual options... just the docs are bust.

gezalore commented on src/astgen:
```diff
@@ -687,6 +816,24 @@ for typen in sorted(Classes.keys()):
+    if "Node" in typen:
+        continue
+    bases = subclasses_of(typen)
+    if "NodeMath" not in bases:
+        continue
+    elif "NodeUniop" in bases:
+        arity = 1
+    elif "NodeBiop" in bases:
+        arity = 2
+    elif "NodeTriop" in bases:
+        arity = 3
+    elif "NodeQuadop" in bases:
+        arity = 4
```
I will think about this, but realistically these are so basic we never change them. I agree, in general it would be nice to have the expression trees be more declarative.

gezalore commented on docs/guide/exe_verilator.rst:
```diff
@@ -344,6 +344,15 @@ Summary:
 
        touch foo.v ; verilator -E --dump-defines foo.v
 
+.. option:: --dump-dfg
+
+   Rarely needed.  Enable writing DFG optimizer .dot files with dumping level 1,
+
+.. option:: --dump-dfgi <level>
+
+   Rarely needed - for developer use.  Set internal DFG dumping level
+   globally to a specific dumping level.
+
```
I am not sure this is an improvement, as I would still have to document the special use of --debugi-<Source>, but more importantly, if you turn this up even moderately high, you will get *a lot* of dot files, which are expensive to output and are almost always noise (unless specifically debugging DFG optimizations), so I don't think you want them if you just do --debug (which sets default level to 3), which your suggested change would do?

wsnyder replied:
I was assuming you would then only dump the files if the debug level was 9.  All across the program if you use debug level 9 you'll may get thousands of dot files.

wsnyder replied:
Perhaps "--dump-treei-V3Dfg 9" would suit us both?  That gives you control, and resolves my worry that we start propagating many debug flags in some non-consistent/extendable way?

wsnyder replied:
(That is there's a filename extension implemented like --debugi- supports.)

gezalore replied:
I appreciate what you are trying to do here. My problem with that proposal is that there are no trees involved here, so --dump-treeei would be a misnomer. Inaccurate naming is one of the things that I think makes code and tools a lot harder to understand and use. If it's a cat then let's not call it a dog and then make people read 200 lines of code to figure out that the reason we feed the dog cat food is because it really is a cat.

As a sidenote: the dump and debug options are already pretty inconsistent (you need various debug levels enabled as the dumps are under `if (debug() >= n)` etc.). The dump-dfg stuff I added in this patch is at least orthogonal to the other switches in that respect.

Btw I'm all for clarity and consistency, but reusing a switch that refers to one internal representation to be used by a different IR is not helping either.

wsnyder replied:
I see your points... Do you have suggestion that won't require we add new flags for the next new fancy optimization with unique dumpers?

gezalore replied:
We could go along the lines you were suggesting, just drop the 'tree' bit? So we have `-dumpi-<thing>` where \<thing\> can be either:
- file basenam like V3Foo, or
- magic word like tree/graph/dfg

wsnyder replied:
Works for me!

gezalore replied:
See #3630

gezalore commented:
I have removed 2 peephole patterns, though these were helping a little (~2%), they were not really peephole in that they looked recursively around in the graph. These can be implemented differently at a later stage.

@wsnyder I would like to merge this so could you please comment?

Of the outstanding issues from above:
- Regarding the debug option: I can do the --debugi-DFG thing if you want, but imho it is clearer as is and would be more helpful for future development to be able to dump those graphs gradually (as in: with multiple dump levels) as it's very noisy.
- Regarding the optimization-off-metacomment: I can re-write the test that depends on this without it. Again I added this mostly so people who cannot share RTL can still bisect their designs without having to edit Verilator itself. We can either just not provide this mechanism, not document the mechanism, or provide an alternative, for example an environment variable based one? (Currently there already are environment variable that can disable individual peephole patterns for debugging/profiling)

wsnyder commented on src/V3Dfg.cpp:
```diff
@@ -0,0 +1,539 @@
+}
+
+static const string toDotId(const DfgVertex& vtx) {
+    string str;
+    str += '"';
```
```suggestion
    string str{'"'};
```

gezalore commented on src/V3DfgDfgToAst.cpp:
```diff
@@ -0,0 +1,313 @@
+    V3DfgOptimizationContext& m_ctx;  // The optimization context for stats
+    AstNodeMath* m_resultp = nullptr;  // The result node of the current traversal
+    // Map from DfgVertex to the AstVar holding the value of that DfgVertex after conversion
+    std::unordered_map<const DfgVertex*, AstVar*> m_resultVars;
```
There is no such thing as DfgVertex::userp at the moment (this is not a V3Graph for various reasons). I was trying to design these passes such that you can run them in parallel on multiple modules, so avoided any global state like the generation number used by the AstNode::user values. I am willing to revise this if this proves to be a performance problem.

gezalore commented on src/V3DfgPeephole.cpp:
```diff
@@ -0,0 +1,1112 @@
+V3DfgPeepholeContext::V3DfgPeepholeContext(const std::string& label)
+    : m_label{label} {
+#define OPTIMIZATION_CHECK_ENABLED(id, name) \
+    m_enabled[id] = V3Os::getenvLongLong("VERILATOR_DFG_PEEPHOLE_ENABLE__" name, 1);
```
This is again just a debug tool really to disable individual optimisations, so I can strip it out. I don't think --debugi-... is appropriate here, we expect --debugi-... Options to print stuff and the i implies an integer argument is expected. The most consistent option based analogue would be -fno-dfg-peephole-<name>

gezalore replied:
Added `-fno-dfg-peephole-<whatnot>`

gezalore commented on src/V3DfgPeephole.cpp:
```diff
@@ -0,0 +1,1112 @@
+                newConcat->lhsp(newLhsp);
+                newConcat->rhsp(newRhsp);
+
+                // Replace this vertex
+                vtxp->replaceWith(newConcat);
+                return true;
```
Can I just add some prints to APPLYING itself (so it prints the name of the pattern and the visited node the pattern applies to)? Some patterns do replacements others just tweak nodes so I don't think there is any other universal information here.

wsnyder replied:
Ok, I just want something printed that I can relatively obviously track down to the code snippet (or one of a couple) likely to be at fault.

gezalore commented on src/V3Dfg.h:
```diff
@@ -0,0 +1,698 @@
+protected:
+    DfgEdge* m_sinksp = nullptr;  // List of sinks of this vertex
+    FileLine* const m_filelinep;  // Source location
+    uint32_t m_width = 0;  // Width of the result of this vertex
```
I could, but:
- I need to pick some canonical packed dtype to represent fixed width types (for example, `logic [31:0]`, `logic [3:0][7:0]`, and `int unsigned` - and even their signed versions - are all the same as far as the DFG IR is concerned, but I need these to be the same otherwise it's intractable to get it right. If you are ok we just use the same packed type for all then this is doable.
- this would mean a back reference to the type table so that needs to be kept live. Currently this is a non issue as we don't collect dead types while a DfgGraph is live.

If you are ok with this I can change  the width to a dtype pointer (but keep a constructor that take a width as well for convenience where appropriate), then add a `width()` method that crashes if the type is not the canonical packed type (I don't want references to unpacked arrays to have a width for example)

wsnyder replied:
I'm ok we use the same packed type for all compatible types.

gezalore replied:
I moved to dtypes. Messes with my parallelisation plan due to the interned types in the type table, but I will sort that out another time.

gezalore commented:
Ok, this is done for now. The final version is better than the numbers atop. Even on OpenTitan it reduces instruction count ~15% (I presume mainly due to better CSE), which translates to about ~11% single threaded speedup there, even with `--trace` (maybe @imphil could proof read for typos :grin:).
